### PR TITLE
Skip non-numeric columns when making a boxplot

### LINKED ISSUE #22799: BUG: boxplot does not work when there is a datetime64 column
Making a boxplot of a DataFrame does not seem to skip datetime64 columns, as you get the following error (it tries to calculate the mean of the datetime64 values):

```
In [4]: df = pd.DataFrame({'a': pd.date_range("2012-01-01", periods=100), 'b': np.random.randn(100), 'c': np.random.randn(100)+2})

In [6]: df.plot(kind='box')
...
TypeError: unsupported operand type(s) for +: 'Timestamp' and 'Timestamp'
```

While when this would be a string column, it would be correctly skipped.

This seems to be a regression, as at least in pandas 0.20.3 this worked.

### PR DESCRIPTION
**Pull request by jrbourbeau**
This PR updates the `BoxPlot` class to skip non-numeric columns when making a DataFrame boxplot. Also raises a more meaningful error message when trying to plot a datetime-like Series.  

Closes #22799.

- [x] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/plotting/_core.py [hunk 1]
```diff
@@ -1597,6 +1598,11 @@ def __init__(self, data, return_type='axes', **kwargs):
                 "return_type must be {None, 'axes', 'dict', 'both'}")
 
         self.return_type = return_type
+        if isinstance(data, ABCDataFrame):
+            data = data.select_dtypes(include='number')
+        elif is_datetimelike(data):
```
Comment (jreback) [inline pandas/plotting/_core.py]: shouldn't this be for any non-DataFrame?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/plotting/_core.py (hunk 1 above)
Comment (jrbourbeau) [reply pandas/plotting/_core.py]: Hmm, I don't follow - I must be missing something. My thinking is that 

```python
if isinstance(data, ABCDataFrame):
    data = data.select_dtypes(include='number')
```

will skip plotting any non-numerical, datetime-like columns, and 

```python
elif is_datetimelike(data):
    raise ValueError(...)
```

will catch any non-DataFrames (e.g. `Series`, `DatetimeIndex`, etc.) with datetime-like data where making a `boxplot` doesn't make sense. 

Is this not the case?

Comment (WillAyd) [reply pandas/plotting/_core.py]: Is this condition specific to box plots? I understand your thinking but it's somewhat orthogonal to the original issue. Might be better served as a separate issue with a more comprehensive change unless it is specific to this type of plot

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
@jrbourbeau can you merge master and address comments?

**Comment by jrbourbeau** [comment]
Thanks for the feedback @WillAyd 

> Is this condition specific to box plots?

Looking at `pandas/plotting/_core.py`, there are instances where datetimes aren't supported in other plots. For instance, in `hist_frame`

https://github.com/pandas-dev/pandas/blob/4b5f4d1a59a7fce09c47eb0c03ba99a6cea472c6/pandas/plotting/_core.py#L2395

and plots that inherit from `PlanePlot` (i.e. `ScatterPlot` and `HexBinPlot`)

https://github.com/pandas-dev/pandas/blob/4b5f4d1a59a7fce09c47eb0c03ba99a6cea472c6/pandas/plotting/_core.py#L800-L803

However, I think these cases are handled on an individual basis. 


I must have missed the use of the `_get_numeric_data()` method when I originally made this PR. I could replace the changes I've made in `BoxPlot` with `data = data._get_numeric_data()` to be more consistent with the existing code. 


> Might be better served as a separate issue with a more comprehensive change

That could be, I'll leave that decision to those with a better understanding of the pandas internals, like yourself

**Comment by WillAyd** [comment]
> I could replace the changes I've made in `BoxPlot` with `data = data._get_numeric_data()` to be more consistent with the existing code.

Go ahead and do this. I'll backtrack on my comment around the error message as I think it's OK to keep in

File: pandas/tests/plotting/test_series.py [hunk 2]
```diff
@@ -675,6 +675,12 @@ def test_boxplot_series(self):
         ylabels = ax.get_yticklabels()
         self._check_text_labels(ylabels, [''] * len(ylabels))
 
+    @pytest.mark.slow
+    def test_boxplot_datetime(self):
```
Comment (WillAyd) [inline pandas/tests/plotting/test_series.py]: Append _raises to the name here

File: pandas/tests/plotting/test_series.py [hunk 3]
```diff
@@ -675,6 +675,12 @@ def test_boxplot_series(self):
         ylabels = ax.get_yticklabels()
         self._check_text_labels(ylabels, [''] * len(ylabels))
 
+    @pytest.mark.slow
```
Comment (WillAyd) [inline pandas/tests/plotting/test_series.py]: Is this actually slow?

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@jrbourbeau can you merge master and address comments

**Comment by WillAyd** [comment]
Closing as stale. Ping if you'd like to continue

**Pull request closed without merging**